TrkHitReco/src/Chi2Clusterer.cc

```diff
@@ -0,0 +1,180 @@
+#include "Offline/TrkHitReco/inc/Chi2Clusterer.hh"
+#include <algorithm>
+#include <vector>
+#include <queue>
+
+namespace mu2e
+{
+  Chi2Clusterer::Chi2Clusterer(const Config& config) :
+    dhit_             (config.hitDistance()),
+    dseed_            (config.seedDistance()),
+    minClusterHits_   (config.minClusterHits()),
+    maxHitdt_         (config.maxHitTimeDiff()),
+    maxNiter_         (config.maxCluIterations()),
+    bkgmask_          (config.bkgmsk()),
+    sigmask_          (config.sigmsk()),
+    testflag_         (config.testflag()),
+    diag_             (config.diag())
+  {
+
+    tbin_     = 1.0;
+    chi2Cut_  = std::pow(((dseed_+dhit_)/2),2);
+    distMethodFlag_ = BkgCluster::chi2;
+
+  }
+
+  //---------------------------------------------------------------------------------------
+  void Chi2Clusterer::init() {}
+
+
+  //----------------------------------------------------------------------------------------------------------
+  void Chi2Clusterer::findClusters(BkgClusterCollection& clusters, const ComboHitCollection& chcol, int iev)
+  {
+
+    std::vector<Chi2BkgHit> BkgHits;
+    BkgHits.reserve(chcol.size());
+    if (chcol.empty()) return;
+
+    initClustering(chcol, BkgHits);
+    doClustering(chcol, clusters, BkgHits);
+
+    long unsigned int minchits = minClusterHits_;
+    auto removePred = [minchits](auto& cluster) {return cluster.hits().size() < minchits;};
+    clusters.erase(std::remove_if(clusters.begin(),clusters.end(),removePred),clusters.end());
+
+    auto transPred = [&BkgHits] (const int i) {return BkgHits[i].chidx_;};
+    for (auto& cluster: clusters) std::transform(cluster.hits().begin(),cluster.hits().end(),cluster.hits().begin(),transPred);
+
+    if(diag_>1) dump(clusters,BkgHits);
+
+  }
+
+
+  //----------------------------------------------------------------------------------------------------------------------
+  void Chi2Clusterer::initClustering(const ComboHitCollection& chcol, std::vector<Chi2BkgHit>& BkgHits)
+  {
+     float maxTime(0);
+     for (size_t ich=0;ich<chcol.size();++ich) {
+       if (testflag_ && (!chcol[ich].flag().hasAllProperties(sigmask_) || chcol[ich].flag().hasAnyProperty(bkgmask_))) continue;
+       BkgHits.emplace_back(Chi2BkgHit(ich));
+       maxTime = std::max(maxTime,chcol[ich].correctedTime());
+     }
+     tbin_ = (maxTime+1.0)/float(numBuckets_);
```

Comment:
It seems odd that the time bin depends on the range.  I would think it should be set according to the resolution.   At least there should be a minimum value corresponding to the resolution, it makes no sense to bin finer than that.  Also, it seems the binning implicitly starts from 0, while there is a minimum hit time due to the flash blanking, so the first bins will always be empty, making the storage inefficient.  I suspect this algorithm was copied from TNTclusterer, which then has the same inefficiency.  The fix is to add a tmin_ variable, set as the minimum hit time, and to subtract that from the hit times when clustering.

Reply:
This is remnant from TNTClusterer, we can discuss it further.